ingest/lambdas/bitstreams-downloader.ts

```diff
@@ -0,0 +1,78 @@
+import { z } from 'zod';
+
+import { s3 } from '../../lib/aws-clients';
+import * as dspaceClient from '../../lib/dspace-client';
+import { S3ObjectLocation, safeGetObjectJson } from '../../lib/s3-utils';
+import lambdaClient from '../../lib/lambda-client';
+import { dspaceItemSchema } from '../../lib/dspace-schemas';
+import { getStringFromEnv } from '../../lib/env-utils';
+import { findPDFBitstreamItem } from '../../lib/dspace-item';
+
+const s3EventSchema = z.object({
+  Records: z.array(
+    z.object({
+      s3: z.object({
+        bucket: z.object({
+          name: z.string().min(1),
+        }),
+        object: z.object({
+          key: z.string().min(1),
+        }),
+      }),
+    })
+  ).nonempty(),
+});
+
+export const handler = async (event: unknown) => {
+  const s3Event = s3EventSchema.parse(event);
+
+  const dspaceEndpoint = getStringFromEnv('DSPACE_ENDPOINT');
+  const bitstreamSourceBucket = getStringFromEnv('DOCUMENT_BINARY_BUCKET');
+  const indexerFunction = getStringFromEnv('INDEXER_FUNCTION_NAME');
+
+  try {
+    // Get the DSpace item from S3.
+    const s3Location = new S3ObjectLocation(
+      s3Event.Records[0].s3.bucket.name,
+      s3Event.Records[0].s3.object.key
+    );
+
+    const dspaceItem = await safeGetObjectJson(
+      s3Location,
+      dspaceItemSchema
+    );
+
+    // Get the PDF bitstream metadata.
+    const pdfBitstreamItem = findPDFBitstreamItem(dspaceItem.bitstreams);
+    if (pdfBitstreamItem !== undefined) {
+      console.log(`INFO: Found PDF for DSpace item ${dspaceItem.uuid}. Uploading to S3.`);
+
+      // Get the PDF from DSpace.
+      const pdfBuffer = await dspaceClient.getBitstream(
+        dspaceEndpoint,
+        pdfBitstreamItem.retrieveLink
+      );
+
+      // Upload the PDF to S3.
+      await s3().putObject({
```

Comment:
How big could these be? Do we want to stream it rather than read it into memory?

Reply:
Can we discuss these two comments tomorrow?

Reply:
I tried the "simple" solution here of using got's stream API but received the error `Non-file stream objects are not supported with SigV4`. Could we pair on this so that we can get it done quickly?

Reply:
Here's an example that you should be able to adapt to this situation:

```ts
import got from 'got';
import { promisify } from 'node:util';
import stream from 'node:stream';
import { s3 } from './aws-clients';
import { createBucket, deleteBucket } from './s3-utils';

const pipeline = promisify(stream.pipeline);

(async () => {
  await deleteBucket('test-bucket', true).catch(() => {});

  await createBucket('test-bucket');

  const pass = new stream.PassThrough();

  s3().upload({
    Bucket: 'test-bucket',
    Key: 'test-key',
    Body: pass
  }).send();

  await pipeline(got.stream('http://www.example.com'), pass);
})();
```